### ISSUE TITLE
engine.state.iteration in distributed training

### ISSUE THREAD

**Issue by dakey860831** [pytorch/ignite#2556]
## ❓ Questions/Help/Support

Hi, I wonder how `engine.state.iteration` update is handled in distributed training? i.e. is it updated globally, or only w.r.t. a local rank. For example, I'd like to log loss for each global iteration from multiple nodes and/or GPUs in a single tensorboard scalar plot; I also would like to add a validation handler triggered by every N global iterations. I looked into `distributed` folder, but didn't find any code explicitly handled this (but I could've total miss that!).

**Comment by vfdev-5** [comment]
@dakey860831 if an `engine` is defined and run by all procs, then `engine.state.iteration` is updated in the same way by all procs.

> For example, I'd like to log loss for each global iteration from multiple nodes and/or GPUs in a single tensorboard scalar plot;
>  I also would like to add a validation handler triggered by every N global iterations.

Here is an example code snippet:
```python
import random
import ignite.distributed as idist
from ignite.engine import Engine, Events


def training(local_rank, *args, **kwargs):

    def train_step(engine, batch):
        # ...
        # forward/backward passes, compute batch loss per proc
        # ...
        batch_loss_per_proc = (20.0 - batch) * 0.1 + random.random()
        return batch_loss_per_proc

    trainer = Engine(train_step)

    # Log batch_loss average across all processes
    @trainer.on(Events.ITERATION_COMPLETED)
    def log_avg_batch_loss():
        # trainer.state.output == batch_loss_per_proc from train_step
        avg_batch_loss = idist.all_reduce(trainer.state.output) / idist.get_world_size()
        if idist.get_rank() == 0:
            print(trainer.state.iteration, "Average batch loss:", avg_batch_loss)

    
    # Validation engine setup:
    def val_step(engine, batch):
        # ...
        y_pred = 0.0 # <--- typically model(batch[0])
        y_true = 0.0 # <--- typically batch[1]
        return y_pred, y_true
        
    evaluator = Engine(val_step)

    # Trigger validation every 5 global iterations
    @trainer.on(Events.ITERATION_COMPLETED(every=5))
    def run_validation():
        # model is DDP wrapped
        # val_dataloader has distributed sampler        
        val_data = range(5)
        state = evaluator.run(val_data)
        # state.metrics will contain computed metrics 
        # if metrics were attached
        
    if idist.get_rank() == 0:
        @evaluator.on(Events.COMPLETED)
        def log_val_metrics():
            print(trainer.state.iteration, "--> Logging val metrics")
            pass


    data = [i for i in range(10)]
    trainer.run(data, max_epochs=2)
    
    
with idist.Parallel("nccl", nproc_per_node=2) as p:
    p.run(training)
```

Output on 2 procs:
```
2022-04-21 09:36:21,497 ignite.distributed.launcher.Parallel INFO: Initialized distributed launcher with backend: 'nccl'
2022-04-21 09:36:21,498 ignite.distributed.launcher.Parallel INFO: - Parameters to spawn processes: 
	nproc_per_node: 2
	nnodes: 1
	node_rank: 0
2022-04-21 09:36:21,499 ignite.distributed.launcher.Parallel INFO: Spawn function '<function training at 0x7fdb6e5bf430>' in 2 processes

1 Average batch loss: 2.409778594970703
2 Average batch loss: 2.511381149291992
3 Average batch loss: 2.3281126022338867
4 Average batch loss: 2.0661051273345947
5 Average batch loss: 2.075127601623535
5 --> Logging val metrics
6 Average batch loss: 1.6815365552902222
7 Average batch loss: 1.9044041633605957
8 Average batch loss: 1.696197748184204
9 Average batch loss: 2.075287342071533
10 Average batch loss: 1.496448278427124
10 --> Logging val metrics
11 Average batch loss: 2.134565830230713
12 Average batch loss: 2.603890895843506
13 Average batch loss: 2.3349111080169678
14 Average batch loss: 2.145655870437622
15 Average batch loss: 2.057931900024414
15 --> Logging val metrics
16 Average batch loss: 1.9581470489501953
17 Average batch loss: 2.0702457427978516
18 Average batch loss: 1.500052571296692
19 Average batch loss: 1.7656314373016357
20 Average batch loss: 1.3346748352050781
20 --> Logging val metrics

2022-04-21 09:36:25,051 ignite.distributed.launcher.Parallel INFO: End of run

```

---


> I looked into distributed folder, but didn't find any code explicitly handled this (but I could've total miss that!).

You can take a look at 
- https://pytorch-ignite.ai/tutorials/intermediate/01-cifar10-distributed/
- https://github.com/pytorch/ignite/blob/master/examples/contrib/cifar10/main.py



Please let us know if something is unclear. Thanks

**Comment by vfdev-5** [comment]
Let me close the issue as answered. Feel free to reopen if needed
